xfer/app_client.go

```diff
@@ -0,0 +1,166 @@
+package xfer
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/rpc"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+
+	"github.com/weaveworks/scope/common/sanitize"
+)
+
+// ScopeProbeIDHeader is the header we use to carry the probe's unique ID. The
+// ID is currently set to the probe's hostname. It's designed to deduplicate
+// reports from the same probe to the same receiver, in case the probe is
+// configured to publish to multiple receivers that resolve to the same app.
+const ScopeProbeIDHeader = "X-Scope-Probe-ID"
+
+// Details are some generic details that can be fetched from /api
+type Details struct {
+	ID      string `json:"id"`
+	Version string `json:"version"`
+}
+
+// AppClient is a client to an app.
+type AppClient interface {
+	Details() (Details, error)
+	ControlConnection(handler ControlHandlerFunc)
+	Stop()
+}
+
+type appClient struct {
+	ProbeConfig
+
+	quit     chan struct{}
+	target   string
+	insecure bool
+	client   http.Client
+
+	controlServerCodecMtx sync.Mutex
```

Comment:
This was added so when we stop an appclient, we can stop the associated control server.  As this requires us checking for a current app server, we need a mutex, as its modified concurrently.  Also synchronises access to controlServerCodec and quit.